src/server/middleware/transaction/transaction.go

```diff
@@ -0,0 +1,83 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transaction
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/goharbor/harbor/src/internal/orm"
+	"github.com/goharbor/harbor/src/server/middleware"
+)
+
+// Config defines the config for Transaction middleware.
+type Config struct {
+	// Acceptor defines a function to commit transaction
+	Acceptor func(statusCode int) bool
+}
+
+var (
+	// DefaultConfig default transaction config
+	DefaultConfig = Config{
+		// acceptor which will commit the transaction when status code between 200 to 308 as success
+		Acceptor: func(statusCode int) bool {
+			// statusCode between 200 to 308 as success
+			return statusCode >= http.StatusOK && statusCode < http.StatusBadRequest
+		},
+	}
+
+	errNonSuccess = errors.New("non success status code")
+)
+
+// Middleware middleware which add transaction for the http request with default config
+func Middleware(skippers ...middleware.Skipper) func(http.Handler) http.Handler {
+	return MiddlewareWithConfig(DefaultConfig, skippers...)
+}
+
+// MiddlewareWithConfig middleware which add transaction for the http request with config
+func MiddlewareWithConfig(config Config, skippers ...middleware.Skipper) func(http.Handler) http.Handler {
+	if config.Acceptor == nil {
+		config.Acceptor = DefaultConfig.Acceptor
+	}
+
+	return middleware.New(func(w http.ResponseWriter, r *http.Request, next http.Handler) {
+		h := func(ctx context.Context) error {
+			rec := &statusRecorder{ResponseWriter: w}
+			next.ServeHTTP(rec, r)
+
+			if !config.Acceptor(rec.statusCode) {
+				return errNonSuccess
+			}
+
+			return nil
+		}
+
+		if err := orm.WithTransaction(h)(r.Context()); err != nil && err != errNonSuccess {
+			// 500 response for db error
+			w.WriteHeader(http.StatusInternalServerError)
```

Comment:
The HTTP status code has been set by the handler, the second call for `WriteHeader` doesn't work. Just log the error?

Reply:
There are two different scenarios for this error:
1. Begin the transaction failed, `WriteHeader`works.
2. Commit or rollback the transaction failed, `WriteHeader` does not work, it will be ignore by the ResponseWriter.

En, I will log the error.

Reply:
Got it, could you add some comment to explain it? And we should response with standard API error here

Reply:
The ResponseWriter for the next handler changed to ResponseBuffer, so this problem was resolved.